### PR TITLE
GG-20691 New source of query metrics.

### PR DESCRIPTION
**Pull request by vsisko**

### CODE REVIEW

File: modules/indexing/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryDetailMetrics.java [hunk 1]
```diff
@@ -64,6 +64,12 @@
     /** Sum of execution time of completions time. */
     private long lastStartTime;
 
+    /** Name of schema. */
+    private String schema;
+
+    /** Flag of local query. */
+    private Boolean local;
```
Comment (akuznetsov-gridgain) [inline modules/indexing/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryDetailMetrics.java]: Do we really need object? Why not use primitive boolean?

File: modules/indexing/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryDetailMetrics.java [hunk 2]
```diff
@@ -195,6 +222,11 @@ public long getLastStartTime() {
         avgTime = in.readDouble();
         totalTime = in.readLong();
         lastStartTime = in.readLong();
+
+        if (protoVer > V1) {
+            schema = U.readString(in);
+            local = (Boolean)in.readObject();
```
Comment (akuznetsov-gridgain) [inline modules/indexing/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryDetailMetrics.java]: Can we use primitive?

File: modules/indexing/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryDetailMetricsCollectorTask.java [hunk 3]
```diff
@@ -116,21 +116,31 @@ private static void aggregateMetrics(long since, Map<GridCacheQueryDetailMetrics
...
+                        oldMetrics.executions() + (int)m.executions(),
+                        oldMetrics.completions() + (int)(m.executions() - m.failures()),
+                        oldMetrics.failures() + (int)m.failures(),
+                        Long.min(oldMetrics.minimumTime(), m.minimumTime()),
+                        Long.max(oldMetrics.maximumTime(), m.maximumTime()),
+                        oldMetrics.totalTime(),
+                        Long.max(oldMetrics.lastStartTime(), m.lastStartTime()),
+                        key)
+                    : new GridCacheQueryDetailMetricsAdapter(SQL_FIELDS, m.query(), null,
+                        (int)m.executions(), (int)(m.executions() - m.failures()), (int)m.failures(), m.minimumTime(),
```
Comment (akuznetsov-gridgain) [inline modules/indexing/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryDetailMetricsCollectorTask.java]: Please, refactor to one line per parameter. This GG code style.

File: modules/indexing/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryResetDetailMetricsTask.java [hunk 4]
```diff
@@ -52,14 +53,7 @@ private VisorCacheResetQueryDetailMetricsJob(Void arg, boolean debug) {
...
         @Override protected Void run(Void arg) {
-            for (String cacheName : ignite.cacheNames()) {
-                IgniteCache cache = ignite.cache(cacheName);
-
-                if (cache == null)
-                    throw new IllegalStateException("Failed to find cache for name: " + cacheName);
-
-                cache.resetQueryDetailMetrics();
-            }
+            ((IgniteH2Indexing)ignite.context().query().getIndexing()).runningQueryManager().resetQueryHistoryMetrics();
```
Comment (akuznetsov-gridgain) [inline modules/indexing/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryResetDetailMetricsTask.java]: May be check for query().getIndexing() instanceOf IgniteH2Indexing for safety?

File: modules/indexing/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryDetailMetrics.java [hunk 5]
```diff
@@ -180,6 +205,8 @@ public long getLastStartTime() {
         out.writeDouble(avgTime);
         out.writeLong(totalTime);
         out.writeLong(lastStartTime);
+        U.writeString(out, schema);
+        out.writeObject(local);
```
Comment (nva) [inline modules/indexing/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryDetailMetrics.java]: Why not out.writeBoolean?

File: modules/indexing/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryDetailMetrics.java (hunk 1 above)
Comment (vsisko) [reply modules/indexing/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryDetailMetrics.java]: New fields have been removed. Failed to collect new metrics without compatibility break.

File: modules/indexing/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryDetailMetrics.java (hunk 5 above)
Comment (vsisko) [reply modules/indexing/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryDetailMetrics.java]: New fields have been removed. Failed to collect new metrics without compatibility break.

File: modules/indexing/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryResetDetailMetricsTask.java (hunk 4 above)
Comment (vsisko) [reply modules/indexing/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryResetDetailMetricsTask.java]: Implemented check of indexing class.

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteFeatures.java [hunk 6]
```diff
@@ -64,8 +64,11 @@
     /** Supports tracking update counter for transactions. */
     TX_TRACKING_UPDATE_COUNTER(12),
 
-    /** Support new security processor */
-    IGNITE_SECURITY_PROCESSOR(13);
+    /** Support new security processor. */
+    IGNITE_SECURITY_PROCESSOR(13),
+
+    /** Indexing enabled. */
+    INDEXING(14);
```
Comment (ygerzhedovich) [inline modules/core/src/main/java/org/apache/ignite/internal/IgniteFeatures.java]: code should be reserved on https://ggsystems.atlassian.net/wiki/spaces/GG/pages/1192198276/Community+edition+features+list

File: modules/indexing/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryDetailMetricsCollectorTask.java (hunk 3 above)
Comment (vsisko) [reply modules/indexing/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryDetailMetricsCollectorTask.java]: Refactored.

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteFeatures.java (hunk 6 above)
Comment (vsisko) [reply modules/core/src/main/java/org/apache/ignite/internal/IgniteFeatures.java]: Code reserved.

**Comment by vsisko** [comment]
> In additonal need to add tests

@ygerzhedovich  Where can we locate that test with excluded ignite-indexing library?

**Comment by vsisko** [comment]
@ygerzhedovich Implemented tests to check INDEXING feature in ignite-core and ignite-indexing packages.
org.apache.ignite.internal.product.IndexingFeatureIsNotAvailableTest
org.apache.ignite.internal.product.IndexingFeatureIsAvailableTest

File: modules/indexing/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryDetailMetricsCollectorTask.java [hunk 7]
```diff
@@ -96,13 +101,21 @@ protected VisorCacheQueryDetailMetricsCollectorJob(@Nullable VisorQueryDetailMet
          * @param since Time when metrics were collected last time.
          * @param res Response.
          * @param metrics Metrics.
+         * @param skipQueryMetrics Skip metrics for SQL and SQL_FIELD queries when {@code true}.
```
Comment (akuznetsov-gridgain) [inline modules/indexing/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryDetailMetricsCollectorTask.java]: Lets rename `skipQueryMetrics` to `collectNotSqlMetrics` and update javadoc accordingly.

File: modules/core/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryCacheDetailMetricsCollectorTask.java [hunk 8]
```diff
@@ -37,10 +37,10 @@
 import static org.apache.ignite.internal.processors.cache.GridCacheUtils.isSystemCache;
 
 /**
- * Task to collect cache query metrics.
+ * Collect query detailed metrics from cache context only. For compatibility when node does not have indexing module.
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryCacheDetailMetricsCollectorTask.java]: Lets change javadoc:
This task should be used to collect query metrics when indexing feature not available.

File: modules/core/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryResetCacheDetailMetricsTask.java [hunk 9]
```diff
@@ -23,10 +23,10 @@
 import org.apache.ignite.internal.visor.VisorOneNodeTask;
 
 /**
- * Reset query detail metrics.
+ * Reset query detailed metrics from cache context only. For compatibility when node does not have indexing module.
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryResetCacheDetailMetricsTask.java]: Lets change javadoc:
This task should be used to reset query metrics when indexing feature not available.

File: modules/indexing/src/test/java/org/apache/ignite/internal/processors/query/SqlMetricsByWCSelfTest.java [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+import org.apache.ignite.internal.visor.query.VisorQueryDetailMetrics;
+import org.apache.ignite.internal.visor.query.VisorQueryDetailMetricsCollectorTask;
+import org.apache.ignite.internal.visor.query.VisorQueryDetailMetricsCollectorTaskArg;
+import org.apache.ignite.internal.visor.query.VisorQueryResetDetailMetricsTask;
+import org.junit.Test;
+
+/**
+ * Tests for ignite SQL system views.
+ */
+public class SqlMetricsByWCSelfTest extends AbstractIndexingCommonTest {
```
Comment (akuznetsov-gridgain) [inline modules/indexing/src/test/java/org/apache/ignite/internal/processors/query/SqlMetricsByWCSelfTest.java]: Lets rename this class to: SqlMetricsOnWebConsoleSelfTest

File: modules/core/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryCacheDetailMetricsCollectorTask.java [hunk 11]
```diff
@@ -37,10 +37,10 @@
...
  */
 @GridInternal
-public class VisorQueryDetailMetricsCollectorTask extends VisorMultiNodeTask<VisorQueryDetailMetricsCollectorTaskArg,
+public class VisorQueryCacheDetailMetricsCollectorTask extends VisorMultiNodeTask<VisorQueryDetailMetricsCollectorTaskArg,
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryCacheDetailMetricsCollectorTask.java]: I think it is better to leave old task with old names.
And add new names for new tasks!

File: modules/indexing/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryDetailMetricsHistoryCollectorTask.java [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+         */
+        protected VisorCacheQueryDetailMetricsCollectorJob(@Nullable VisorQueryDetailMetricsCollectorTaskArg arg, boolean debug) {
+            super(arg, debug);
+        }
+
+        /**
+         * @param since Time when metrics were collected last time.
+         * @param res Response.
+         * @param metrics Metrics.
+         * @param collectNotSqlMetrics When {@code true} collect metrics for no SQL queries only.
```
[the remark this answers is not in the document]
Comment (vsisko) [reply modules/indexing/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryDetailMetricsHistoryCollectorTask.java]: Fixed

File: modules/indexing/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryDetailMetricsHistoryCollectorTask.java [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+        for (GridCacheQueryDetailMetricsAdapter m: aggMetrics)
+            res.add(new VisorQueryDetailMetrics(m));
+
+        return res;
+    }
+
+    /**
+     * Job that will actually collect query metrics.
+     */
+    private static class VisorCacheQueryDetailMetricsCollectorJob
```
[the remark this answers is not in the document]
Comment (vsisko) [reply modules/indexing/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryDetailMetricsHistoryCollectorTask.java]: Renamed

File: modules/indexing/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryDetailMetricsHistoryCollectorTask.java [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+
+import static org.apache.ignite.internal.processors.cache.GridCacheUtils.isSystemCache;
+import static org.apache.ignite.internal.processors.cache.query.GridCacheQueryType.SQL;
+import static org.apache.ignite.internal.processors.cache.query.GridCacheQueryType.SQL_FIELDS;
+
+/**
+ * Task to collect cache query metrics.
+ */
+@GridInternal
+public class VisorQueryDetailMetricsHistoryCollectorTask extends VisorMultiNodeTask<VisorQueryDetailMetricsCollectorTaskArg,
```
[the remark this answers is not in the document]
Comment (vsisko) [reply modules/indexing/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryDetailMetricsHistoryCollectorTask.java]: Renamed

File: modules/core/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryCacheDetailMetricsCollectorTask.java (hunk 11 above)
Comment (vsisko) [reply modules/core/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryCacheDetailMetricsCollectorTask.java]: Old task renamed to old name.

File: modules/indexing/src/test/java/org/apache/ignite/internal/processors/query/SqlMetricsByWCSelfTest.java (hunk 10 above)
Comment (vsisko) [reply modules/indexing/src/test/java/org/apache/ignite/internal/processors/query/SqlMetricsByWCSelfTest.java]: Renamed.

**Pull request merged**